src/docs/arch-fast-truncate.dox

```diff
@@ -0,0 +1,847 @@
+and no cell will be produced.
+
+Second, pages that reconcile empty end up deleted.
+During a checkpoint this happens in the checkpoint cleanup; in eviction, it happens in the
+\c WT_PM_REC_EMPTY case of \c __evict_page_dirty_update.
+The ref state becomes \c WT_REF_DELETED and no \c page_del is generated.
+(Also note that \c WT_REF_MEM pages with \c WT_PM_REC_EMPTY reconciliation results are
+explicitly skipped during internal page reconciliation.)
+
+In VLCS, empty deleted pages are inserted during certain internal page reads; see @ref
+ft_vlcs.
+
+Finally, new trees are created with a single empty deleted leaf page, because creating the
+tree with no leaves at all causes problems.
+
+@section ft_page_skip Skipping deleted pages
+
+There are two separate skip functions for skipping deleted pages during tree walks.
+The basic skip function is \c __wt_delete_page_skip in \c bt_delete.c; it is always called
+in tree walks when \c WT_READ_SKIP_DELETED is set, which is all tree walks except RTS.
+It checks for whether the ref state is \c WT_REF_DELETED and the truncation is visible to
+the caller.
+
+As in other cases (see @ref ft_visibility) we must treat prepared truncations as not
+visible; in order to generate prepare conflicts we cannot skip over truncations that are
+prepared but not yet committed.
+
+The other skip function is \c __wt_btcur_skip_page in \c btree_inline.h.
+This is used by cursor next and previous to skip over deleted pages in those traversals
+specifically, and in addition to checking for \c WT_REF_DELETED with visible \c page_del,
+it also inspects the address cell.
+If the page is in memory and unmodified and the address cell contains page-delete
+information, it checks for the visibility of that deletion.
+Failing that, it checks the time aggregate in case all the values on the page are no
+longer visible, e.g. because the page was slow-truncated and reconciled.
+
+(For an unmodified in-memory page the deletion information in the address should be the
+same as present in the ref's \c page_del structure.
+However, inspecting the unpacked copy instead, which is free since we are unpacking the
+cell to look at the time aggregate anyway, makes this check independent of the lifecycle
+of the ref's \c page_del.
+This was more important in previous iterations of the code than it is now; however, in
+principle it's possible to drop the ref's \c page_del immediately upon instantiation in a
+read-only tree, since it is kept for internal page reconciliation and pages in read-only
+trees are never reconciled.
+Checking the unpacked information here avoids interfering with that option.)
+
+Note that \c __wt_btcur_skip_page is passed to the tree-walk code as a custom skip
+function, which means that when it's used <i>both</i> it and \c __wt_delete_page_skip are
+checked.
+This is not optimal and probably ought to be tidied.
+Furthermore, it is unclear whether these functions really need to be different; that is,
```

Comment:
Should we create a ticket for this investigation ?

Reply:
I guess so... WT-9859